### PR TITLE
Implements a fix for ADAL cache replication

### PR DESCRIPTION
**Pull request by iambmelt**
Testing still in progress, but the outline of a fix for the below bug:
- https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1437575

### CODE REVIEW

File: adal/src/main/java/com/microsoft/aad/adal/DelegatingCache.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+import static com.microsoft.aad.adal.TokenCacheAccessor.getMsalOAuth2TokenCache;
+
+import android.content.Context;
+
+import androidx.annotation.NonNull;
+
+import java.util.Iterator;
+
+class DelegatingCache implements ITokenCacheStore {
```
Comment (iambmelt) [inline adal/src/main/java/com/microsoft/aad/adal/DelegatingCache.java:33]: TODO javadoc

File: adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java [hunk 2]
```diff
@@ -98,10 +98,11 @@
         sharedSSOCaches.add(msalOAuth2TokenCache);
         mCommonCache = new ADALOAuth2TokenCache(appContext, sharedSSOCaches);
 
-        if (mTokenCacheStore instanceof DefaultTokenCacheStore) {
+        if (mTokenCacheStore instanceof DelegatingCache) {
+            final ITokenCacheStore delegate = ((DelegatingCache) mTokenCacheStore).getDelegateCache();
             //If the default token cache is in use... delegate token operations to unified cache in common
             //If not using default token cache then sharing SSO state between ADAL & MSAL cache implementations will not be possible anyway
-            mUseCommonCache = true;
+            mUseCommonCache = delegate instanceof DefaultTokenCacheStore;
```
Comment (iambmelt) [inline adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java:105]: Consider using `isAssignableFrom`

File: adal/src/main/java/com/microsoft/aad/adal/DelegatingCache.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    private final Context mContext;
+    private final ITokenCacheStore mDelegate;
+
+    DelegatingCache(@NonNull final Context context,
```
Comment (iambmelt) [inline adal/src/main/java/com/microsoft/aad/adal/DelegatingCache.java:38]: Javadoc

File: adal/src/main/java/com/microsoft/aad/adal/DelegatingCache.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                    @NonNull final ITokenCacheStore delegate) {
+        mContext = context;
+        mDelegate = delegate;
+    }
+
+    public ITokenCacheStore getDelegateCache() {
```
Comment (iambmelt) [inline adal/src/main/java/com/microsoft/aad/adal/DelegatingCache.java:44]: Javadoc

**Review by irving-as — APPROVED** [review]

File: adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java (hunk 2 above)
Comment (AdamBJohnsonx) [reply adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java:105]: Is it possible to write a unit test here that highlights the problem, and then is fixed by this code change?

Comment (iambmelt) [reply adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java:105]: 21b0ce12526b95bb6a803b75b059282e31c6272e

**Comment by iambmelt** [comment]
> This test is detecting a particular kind of problem. Putting a javadoc block explaining what it was helps people reading the test without naming it something like testMsalCacheIsUpdatedWhenAdalCacheIsPresentAndAWriteGoesToTheLatter...
> 
> I think that's what we're doing...

Added javadoc, constants, cleaned up tests

File: adal/src/androidTest/java/com/microsoft/aad/adal/TokenCacheAccessorTests.java [hunk 5]
```diff
@@ -169,14 +209,42 @@ public void setUp() throws Exception {
...
         mTokenCacheAccessor = new TokenCacheAccessor(
-                mockContext,
+                mContext,
                 tokenCacheStore,
                 WORLDWIDE_AUTHORITY,
                 UUID.randomUUID().toString()
         );
+
+
+        // Populate a mock Instance Discovery
```
Comment (AdamBJohnsonx) [inline adal/src/androidTest/java/com/microsoft/aad/adal/TokenCacheAccessorTests.java:222]:
```suggestion
        // Populate a mock Instance Discovery containing entries for the common authority as
        // well as one instance for Mooncake.
```

**Comment by iambmelt** [comment]
One of the ADAL tests is failing... looking into it to see if related to this PR

File: adal/src/androidTest/java/com/microsoft/aad/adal/TokenCacheAccessorTests.java [hunk 6]
```diff
@@ -204,58 +271,130 @@ public void testUpdateTokenCacheUsesResultAuthority() throws MalformedURLExcepti
...
+                TID,
+                MOCK_ID_TOKEN_WITH_CLAIMS,
+                null,
+                CLIENT
         );
 
+        result.setAuthority(WORLDWIDE_AUTHORITY);
+        result.setClientInfo(new ClientInfo(MOCK_CLIENT_INFO));
+        result.setResponseReceived(System.currentTimeMillis());
+        result.setExpiresIn(System.currentTimeMillis());
```
Comment (AdamBJohnsonx) [inline adal/src/androidTest/java/com/microsoft/aad/adal/TokenCacheAccessorTests.java:348]: That's a weird way to set expiresIn..

**Comment by iambmelt** [comment]
> One of the ADAL tests is failing... looking into it to see if related to this PR

Root caused the issue to this constant
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blame/dev/common/src/main/java/com/microsoft/identity/common/internal/cache/ADALOAuth2TokenCache.java#L122

**Comment by iambmelt** [comment]
> > One of the ADAL tests is failing... looking into it to see if related to this PR
> 
> Root caused the issue to this line
> https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blame/dev/common/src/main/java/com/microsoft/identity/common/internal/cache/ADALOAuth2TokenCache.java#L122

Opened up a PR to resolve the issue here:
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1444

**Review by AdamBJohnsonx — APPROVED** [review]
Only comment, is it expiresIn or expiresAt?  Setting expiresIn to the currentTime in milliseconds seems… strange.

**Pull request merged**